analytics/location_heatmaps/geo_utils.py

```diff
@@ -167,37 +201,74 @@ def rebuild_from_vector(vector, tree, image_size, contour=False, threshold=0):
     image_size: desired final resolution of the image.
     contour: release only the contours of the grid (for debugging)
     threshold: reduces noise by setting values below threshold to 0.
+    positivity: produce two images with positive and negative cases.
+    count_min: use count min sketch.
 
   Returns:
     image of the size `image_size x image_size`
   """
   image_bit_level = int(np.log2(image_size))
   current_image = np.zeros([image_size, image_size])
+  pos_image, neg_image = None, None
+  if positivity:
+    pos_image = np.zeros([image_size, image_size])
+    neg_image = np.zeros([image_size, image_size])
   for path in sorted(tree):
-    value = vector[tree[path]]
-    (x, y, prefix_len) = binary_path_to_coordinates(path)
+    if count_min:
+      value = count_min.query(path)
+    else:
+      value = vector[tree[path]]
+    (x, y, prefix_len, pos) = binary_path_to_coordinates(path)
     (x_bot, x_top, y_bot,
-     y_top) = transform_region_to_coordinates(x, y, prefix_len, image_bit_level)
+     y_top) = transform_region_to_coordinates(x, y, prefix_len,
+                                              image_bit_level)
+
     if value < threshold:
       value = 0
-    count = value / 2**(2 * (image_bit_level - prefix_len))
+    count = value / 2 ** (1 * (image_bit_level - prefix_len))
 
     # Build a grid image without filling the regions.
     if contour:
-      current_image[x_bot:x_top + 1, y_bot:y_bot + 1] += 1
-      current_image[x_bot:x_top + 1, y_top:y_top + 1] += 1
-      current_image[x_bot:x_bot + 1, y_bot:y_top + 1] += 1
-      current_image[x_top:x_top + 1, y_bot:y_top + 1] += 1
+      current_image[x_bot:x_top + 1,
+      y_bot - max(1, 5 // prefix_len):y_bot + max(1, 5 // prefix_len)] = 1
+      current_image[x_bot:x_top + 1,
+      y_top - max(1, 5 // prefix_len):y_top + 10 // prefix_len] = 1
+      current_image[
+      x_bot - max(1, 5 // prefix_len):x_bot + 10 // prefix_len,
+      y_bot:y_top + 1] = 1
+      current_image[
+      x_top - max(1, 5 // prefix_len):x_top + 10 // prefix_len,
+      y_bot:y_top + 1] = 1
     else:
+      # balance for collapsing
+      depth = len(path.split('/'))
+      keys = tree.keys(path)[1:]
+      if len(keys) > 1:
+        for key in keys:
+          sub_node_depth = len(key.split('/')) - depth
+          scale = 4**sub_node_depth/ (4**sub_node_depth-1)
+          count *= scale
+
       current_image[x_bot:x_top + 1, y_bot:y_top + 1] = count
-  return current_image
+      if positivity:
+        if pos == 1:
+          pos_image[x_bot:x_top + 1, y_bot:y_top + 1] = count
+        elif pos == 0:
+          neg_image[x_bot:x_top + 1, y_bot:y_top + 1] = count
+        else:
+          raise ValueError(f'Not supported: {pos}')
+  return current_image, pos_image, neg_image
 
 
 def split_regions(tree_prefix_list,
```

Comment:
this function is very long with a lot of nested conditions, which makes it hard to read. can we extract some reasonable helpers to improve readability by making the overall structure of the threshold checks and tree traversal more apparent?

Reply:
done